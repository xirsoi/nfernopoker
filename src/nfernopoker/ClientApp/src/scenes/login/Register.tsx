import * as React from "react";
import { Component, ChangeEvent } from 'react';
import { compose } from "redux";
import { connect } from "react-redux";
import { withFirebase } from "react-redux-firebase";
import { withRouter, RouteComponentProps } from "react-router";
import { Button, TextField, CardContent, CardActions, Typography } from "@material-ui/core";
import { sendMessageAction } from "./actions";

interface ILocalProps {
  classes: any;
  secondaryButtonText: string;
  onSecondaryButton: () => void;
}
interface ILocalState {
  firstName: string;
  lastName: string;
  email: string;
  password: string;
}
interface IFirebase {
  firebase: any;
}
interface IConnectedDispatch {
  sendMessage: (message: string) => void
}
type IProps = ILocalProps & IConnectedDispatch & RouteComponentProps<any> & IFirebase;

class RegisterComponent extends Component<IProps, ILocalState> {

  constructor(
    public props: IProps
  ) {
    super(props);

    this.state = {
      firstName: '',
      lastName: '',
      email: '',
      password: ''
    };
  }

  public handleFormChanged(event: ChangeEvent<HTMLInputElement>, name: string): void {
    let newState = { ...this.state };
    newState[name] = event.target.value;
    this.setState(newState);
  }

  public handleSubmit = (event: ChangeEvent<HTMLInputElement>) => {
    event.preventDefault();
    event.stopPropagation();
    this.props.firebase.createUser(
      { email: this.state.email, password: this.state.password },
      { firstName: this.state.firstName, lastName: this.state.lastName, email: this.state.email }
    ).then((r: any) => {
      this.props.history.push("/games");
    }).catch((ex: any) => this.props.sendMessage(ex.message));
  }

  public render() {
    let isDisabled = !(this.state.firstName && this.state.lastName && this.state.email && this.state.password);
    return (
      <React.Fragment>
        <form id="register-form" onSubmit={(event: any) => this.handleSubmit(event)}>
          <CardContent>
            <Typography className={this.props.classes.title} color="textSecondary">
              Create a new account
          </Typography>
            <TextField
              id="first-name"
              required={true}
              fullWidth={true}
              label="First Name"
              onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "firstName")}
              className={this.props.classes.button}
            />
            <TextField
              id="last-name"
              required={true}
              fullWidth={true}
              label="Last Name"
              onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "lastName")}
              className={this.props.classes.button}
            />
            <TextField
              id="email"
              required={true}
              fullWidth={true}
              type="email"
              label="Email"
              onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "email")}
              className={this.props.classes.button}
            />
            <TextField
              id="password"
              required={true}
              fullWidth={true}
              type="password"
              label="Password"
              onChange={(event: ChangeEvent<HTMLInputElement>) => this.handleFormChanged(event, "password")}
              className={this.props.classes.button}
            />
          </CardContent>
          <CardActions>
            <Button
              id="submit"
              form="register-form"
              type="submit"
              className={this.props.classes.button}
              disabled={isDisabled}
              variant="contained"
              style={{ marginLeft: '16px' }}
              title="Register"
              color="primary">
              Register
            </Button>
            <Button
              id="secondary-btn"
              onClick={this.props.onSecondaryButton}
              size="small">
              {this.props.secondaryButtonText}
            </Button>
          </CardActions>
        </form>
      </React.Fragment>
    );
  }
}

export const RegisterTestComp = RegisterComponent;

export const Register: React.ComponentClass<ILocalProps> = compose<React.ComponentClass<ILocalProps>>(
  connect(null, { sendMessage: sendMessageAction }),
  withRouter,
  withFirebase
)(RegisterComponent)
