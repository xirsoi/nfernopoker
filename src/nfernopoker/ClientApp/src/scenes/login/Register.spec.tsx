import * as React from "react";
import { shallow, ShallowWrapper } from "enzyme";

import * as enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { RegisterTestComp } from "./Register";
import { CardContent } from "@material-ui/core";

enzyme.configure({ adapter: new Adapter() })

let component: ShallowWrapper<any, any, any>;

const fakeProps = {
  classes: {},
  secondaryButtonText: "Login instead",
  onSecondaryButton: jest.fn(),
  sendMessage: jest.fn(),
  firebase: {
    createUser: jest.fn().mockResolvedValue({})
  },
  history: ({ push: jest.fn() } as any),
  match: (jest.fn() as any),
  location: (jest.fn() as any),
  params: (jest.fn() as any)
};

describe("Register Component", () => {

  beforeEach(() => {
    component = shallow(<RegisterTestComp {...fakeProps} />);
  });

  it("should render without error", () => {
    expect(component.instance()).toMatchSnapshot();
  });

  it("should always render card content", () => {
    const divs = component.find(CardContent);
    expect(divs.length).toBeGreaterThan(0);
  });


  describe("when secondary button is clicked", () => {

    beforeEach(() => {
      component.find("#secondary-btn").simulate("click");
    });

    it("should call the callback fuction", () => {
      expect(fakeProps.onSecondaryButton).toHaveBeenCalled();
    });

  });

  describe("when valid form was filled", () => {

    beforeEach(() => {
      component.find("#first-name").simulate("change", { target: { name: "first-name", value: "Bob" } });
      component.find("#last-name").simulate("change", { target: { name: "first-name", value: "Smith" } });
      component.find("#email").simulate("change", { target: { name: "email", value: "blah@gmail.com" } });
      component.find("#password").simulate("change", { target: { name: "password", value: "pword123" } });
    });

    describe("and when form was submitted successfully", () => {

      beforeEach(() => {
        const fakeEvent = {
          preventDefault: () => jest.fn(),
          stopPropagation: () => jest.fn()
        };
        component.find("#register-form").simulate("submit", fakeEvent);
      });

      it("should create a new user in firebase", () => {
        expect(fakeProps.firebase.createUser).toHaveBeenCalledWith(
          { email: "blah@gmail.com", password: "pword123" },
          { firstName: "Bob", lastName: "Smith", email: "blah@gmail.com" }
        );
      });

      it("should change url to go to games", () => {
        expect(fakeProps.history.push).toHaveBeenCalledWith("/games");
      });

    });

    describe("and when form submit errors", () => {
      const messageObj = { message: "It all went wrong Bob" };
      beforeEach(() => {

        fakeProps.firebase.createUser = jest.fn().mockRejectedValue(messageObj);
        const fakeEvent = {
          preventDefault: () => jest.fn(),
          stopPropagation: () => jest.fn()
        };
        component.find("#register-form").simulate("submit", fakeEvent);
      });

      it("should send a message so the user is notified", () => {
        expect(fakeProps.sendMessage).toHaveBeenCalledWith(messageObj.message);
      });

    });

  });

  describe("when form was incompletly filled", () => {

    beforeEach(() => {
      component.find("#first-name").simulate("change", { target: { name: "first-name", value: "Bob" } });
    });

    it("should have a disabled submit button", () => {
      const btn = component.find("#submit");
      expect(btn.prop("disabled")).toBeTruthy();
    });

  });

});
