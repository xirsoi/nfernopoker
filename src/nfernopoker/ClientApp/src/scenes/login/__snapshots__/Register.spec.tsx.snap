// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component should render without error 1`] = `
RegisterComponent {
  "context": Object {},
  "handleSubmit": [Function],
  "props": Object {
    "classes": Object {},
    "firebase": Object {
      "createUser": [MockFunction],
    },
    "history": Object {
      "push": [MockFunction],
    },
    "location": [MockFunction],
    "match": [MockFunction],
    "onSecondaryButton": [MockFunction],
    "params": [MockFunction],
    "secondaryButtonText": "Login instead",
    "sendMessage": [MockFunction],
  },
  "refs": Object {},
  "setState": [Function],
  "state": Object {
    "email": "",
    "firstName": "",
    "lastName": "",
    "password": "",
  },
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_element": <RegisterComponent
        classes={Object {}}
        firebase={
          Object {
            "createUser": [MockFunction],
          }
        }
        history={
          Object {
            "push": [MockFunction],
          }
        }
        location={[MockFunction]}
        match={[MockFunction]}
        onSecondaryButton={[MockFunction]}
        params={[MockFunction]}
        secondaryButtonText="Login instead"
        sendMessage={[MockFunction]}
      />,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_newState": null,
      "_rendered": <React.Fragment>
        <form
          id="register-form"
          onSubmit={[Function]}
        >
          <WithStyles(CardContent)>
            <WithStyles(Typography)
              color="textSecondary"
            >
              Create a new account
            </WithStyles(Typography)>
            <TextField
              fullWidth={true}
              id="first-name"
              label="First Name"
              onChange={[Function]}
              required={true}
              select={false}
              variant="standard"
            />
            <TextField
              fullWidth={true}
              id="last-name"
              label="Last Name"
              onChange={[Function]}
              required={true}
              select={false}
              variant="standard"
            />
            <TextField
              fullWidth={true}
              id="email"
              label="Email"
              onChange={[Function]}
              required={true}
              select={false}
              type="email"
              variant="standard"
            />
            <TextField
              fullWidth={true}
              id="password"
              label="Password"
              onChange={[Function]}
              required={true}
              select={false}
              type="password"
              variant="standard"
            />
          </WithStyles(CardContent)>
          <WithStyles(CardActions)>
            <WithStyles(Button)
              color="primary"
              disabled={true}
              form="register-form"
              id="submit"
              style={
                Object {
                  "marginLeft": "16px",
                }
              }
              title="Register"
              type="submit"
              variant="contained"
            >
              Register
            </WithStyles(Button)>
            <WithStyles(Button)
              id="secondary-btn"
              onClick={[MockFunction]}
              size="small"
            >
              Login instead
            </WithStyles(Button)>
          </WithStyles(CardActions)>
        </form>
      </React.Fragment>,
      "_rendering": false,
      "_updater": [Circular],
    },
  },
  Symbol(enzyme.__setState__): [Function],
}
`;
